### ISSUE TITLE
Parsing by sh is broken

### ISSUE THREAD

**Issue by TheLocehiliosan** [yadm-dev/yadm#299]
### Describe the bug

```
$ sh /usr/local/bin/yadm
/usr/local/bin/yadm: line 1337: syntax error near unexpected token `<'
/usr/local/bin/yadm: line 1337: `      done < <("$GIT_PROGRAM" -C "$YADM_WORK" submodule status)'
```

### Additional context

For portability, yadm is initially run via `/bin/sh` (which has been the most portable shebang location), and then subsequently run using "bash" found in the PATH. Recent changes to upgrade has broken the parsing by "sh".

**Comment by TheLocehiliosan** [comment]
This has been fixed in 3.0.2, but a test still needs to be added to prevent this in the future.

**Comment by erijo** [comment]
Have you considered using `#!/usr/bin/env bash` instead?

**Comment by TheLocehiliosan** [comment]
Yes, it turns out the location and availability of `env` is less reliable than the existence of `/bin/sh`. 😟

**Comment by TheLocehiliosan** [comment]
It seems this problem has less to do with being able to parse all of yadm with `sh` (has failed full syntax parsing of yadm for quite some time, and still does). Usually `sh` only parses the very beginning of yadm, until it does an exec on bash.

This problem was really on if MacOS `/bin/sh` could process yadm. It seems process substitution breaks this. I'm not sure how to best test this in Docker just yet.

**Comment by erijo** [comment]
I think the problem here is that `/bin/sh` on mac is bash in posix mode.

If I change the start of yadm to this, 3.0.1 works for me when I test on a mac (I've only tested `yadm version`):
```sh
if [ -z "$BASH_VERSION" ]; then
  [ "$YADM_TEST" != 1 ] && exec bash "$0" "$@"
fi

[[ -o posix ]] && set +o posix
```

But as the size of /bin/sh and /bin/bash on mac differs it might be better to do: `[[ -o posix ]] && exec bash +o posix "$0" "$@"`. I don't think `+o posix` is actually needed, but I added it just to be sure.

**Comment by TheLocehiliosan** [comment]
Yes. Yesterday I arrived at the same assessment, that `set +o posix` would allow the process substitution to be included. However, I'm undecided on how I want to proceed. I think I'm choosing between two options.

* The first is to `set +o posix` (after confirming running within a version of Bash).

* The second option would be to come up with a feasible way to test with the MacOS provided Bash 3, and using those tests to avoid features that break it. (the problem isn't merely `-o posix`, as other versions of Bash are quite happy with yadm 3.0.1 with `set -o posix`)

The first option is my clear favorite, but the only thing causing me pause is doing `set +o posix` is a definitive step away from ever supporting all POSIX-compliant shells.

I'm not particularly enthusiastic about going down the POSIX-compliant road anyway. _It is already painful enough to support the older version of Bash, abstaining from associative arrays which would make the alternate scoring so much easier to read and maintain. I do think POSIX-compliant would be a new level of work-arounds that would be even more difficult to maintain._

Your comment does prompt a question for me. Why do you suggest `[[ -o posix ]] && set +o posix` instead of just `set +o posix`? Is there an advantage to doing it conditionally?

**Comment by erijo** [comment]
I think I just started with the condition and then never thought about removing it.

The only issue I see with only doing the set is if sh differs from bash in some more way on mac. At least on the system I tried they are different binaries with very big size difference. So it might be good to re-exec with bash.

Btw, does anyone know how long bash will be available on mac?

**Comment by erijo** [comment]
Tests are as of bb21c9a26 run directly on macOS, using the system bash. I did a test where I reverted [b32025b](https://github.com/yadm-dev/yadm/commit/b32025bcc3c43008ead291f754d28d66c9035319) and the build then [fails as expected](https://github.com/yadm-dev/yadm/actions/runs/12999484810). So based on this I'm closing this issue.
